test/e2e_node/restart_test.go

```diff
@@ -118,5 +142,93 @@ var _ = framework.KubeDescribe("Restart [Serial] [Slow] [Disruptive]", func() {
 				By(fmt.Sprintf("Docker restart test passed with %d pods", len(postRestartRunningPods)))
 			})
 		})
+		Context("Live restore", func() {
+
+			It("should restore container when live restore is on", func() {
+				framework.SkipIfContainerRuntimeIs("rkt")
+
+				// Skip when docker engine version is less than 1.12
+				dockerVersion, err := getDockerVersion()
+				dockerRegex, _ := regexp.Compile(`1\.12\.[0-9]+`)
+				if !dockerRegex.Match([]byte(dockerVersion)) {
+					framework.Skipf("Live restore isn't suported before Docker Engine version 1.12")
+				}
+
+				// skip when docker live-restore is not enabled
+				dockerPid, err := ioutil.ReadFile("/var/run/docker.pid")
+				framework.ExpectNoError(err, "Failed to get docker daemon PID: %v, stdout: %q", err, string(dockerPid))
+
+				dockerdCmd, _ := ioutil.ReadFile("/proc/" + string(dockerPid) + "/cmdline")
+
+				daemonConfFile := "/etc/docker/daemon.json"
+				_, err = os.Stat(daemonConfFile)
+				var inConfFile bool
+				if !os.IsNotExist(err) {
+					opt := new(DockerDaemonOption)
+					data, err := ioutil.ReadFile(daemonConfFile)
+					err = json.Unmarshal(data, &opt)
+					framework.ExpectNoError(err, "Failed to parse docker config file %q: %v", daemonConfFile, err)
+					inConfFile = opt.Live_restore
+				}
+				asOption := strings.Contains(string(dockerdCmd), "--live-restore")
+
+				if !asOption && !inConfFile {
+					framework.Skipf("Docker live restore is not set")
+				}
+
+				// create pod
+				podClient := f.PodClient()
+				name := "pod-test-live-restore"
+				value := strconv.Itoa(time.Now().Nanosecond())
+				pod := &v1.Pod{
+					ObjectMeta: apimetav1.ObjectMeta{
+						Name: name,
+						Labels: map[string]string{
+							"name": "foo",
+							"time": value,
+						},
+					},
+					Spec: v1.PodSpec{
+						Containers: []v1.Container{
+							{
+								Name:  "nginx",
+								Image: "gcr.io/google_containers/nginx-slim:0.7",
+							},
+						},
+					},
+				}
+
+				preRestartPod := podClient.CreateSync(pod)
+				By(fmt.Sprintf("Pre restart container id %q", preRestartPod.Status.ContainerStatuses[0].ContainerID))
+
+				// restart docker
+				if stdout, err := exec.Command("sudo", "systemctl", "restart", "docker").CombinedOutput(); err != nil {
```

Comment:
i feel like we should check if we are on systemd instead of trying to call the command directly...
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/services/kubelet.go#L104

can we move logic for docker start/restart into test/services/docker.go or something so if we have other tests, we can have shared logic?